applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp

```diff
@@ -371,187 +340,260 @@ namespace Kratos
-                CriticalHeadFile.close();
+    void KratosExecute::AddNodalSolutionStepVariables(ModelPart& model_part) const
+    {
+        model_part.AddNodalSolutionStepVariable(VELOCITY);
+        model_part.AddNodalSolutionStepVariable(ACCELERATION);
+
+        // Displacement
+        model_part.AddNodalSolutionStepVariable(DISPLACEMENT);
+        model_part.AddNodalSolutionStepVariable(TOTAL_DISPLACEMENT);
+        model_part.AddNodalSolutionStepVariable(REACTION);
+        model_part.AddNodalSolutionStepVariable(POINT_LOAD);
+        model_part.AddNodalSolutionStepVariable(LINE_LOAD);
+        model_part.AddNodalSolutionStepVariable(SURFACE_LOAD);
+        model_part.AddNodalSolutionStepVariable(VOLUME_ACCELERATION);
+        model_part.AddNodalSolutionStepVariable(NORMAL_CONTACT_STRESS);
+        model_part.AddNodalSolutionStepVariable(TANGENTIAL_CONTACT_STRESS);
+
+        // Water
+        model_part.AddNodalSolutionStepVariable(WATER_PRESSURE);
+        model_part.AddNodalSolutionStepVariable(REACTION_WATER_PRESSURE);
+        model_part.AddNodalSolutionStepVariable(DT_WATER_PRESSURE);
+        model_part.AddNodalSolutionStepVariable(NORMAL_FLUID_FLUX);
+        model_part.AddNodalSolutionStepVariable(HYDRAULIC_DISCHARGE);
+
+        // Smoothing
+        model_part.AddNodalSolutionStepVariable(NODAL_AREA);
+        model_part.AddNodalSolutionStepVariable(NODAL_CAUCHY_STRESS_TENSOR);
+        model_part.AddNodalSolutionStepVariable(NODAL_DAMAGE_VARIABLE);
+        model_part.AddNodalSolutionStepVariable(NODAL_JOINT_AREA);
+        model_part.AddNodalSolutionStepVariable(NODAL_JOINT_WIDTH);
+        model_part.AddNodalSolutionStepVariable(NODAL_JOINT_DAMAGE);
+    }
+
+    int KratosExecute::FindCriticalHead(ModelPart& model_part,
+                                        const Kratos::Parameters& gid_output_settings,
+                                        const CriticalHeadInfo& criticalHeadInfo,
+                                        LoggerOutput::Pointer p_output,
+                                        const shared_ptr<Process>& RiverBoundary,
+                                        const GeoMechanicsNewtonRaphsonErosionProcessStrategyType::Pointer& p_solving_strategy,
+                                        const CallBackFunctions& rCallBackFunctions)
+    {
 
-                KRATOS_INFO_IF("GeoFlowKernel", this->GetEchoLevel() > 0) << "Finished writing result" << std::endl;
+        auto currentProcess = std::static_pointer_cast<GeoFlowApplyConstantHydrostaticPressureProcess>(RiverBoundary);
+        KRATOS_INFO_IF("GeoFlowKernel", this->GetEchoLevel() > 0) << "River boundary name: " << currentProcess->GetName() << std::endl;
+
+        currentProcess->SetReferenceCoord(criticalHeadInfo.minCriticalHead);
+        mCurrentHead = criticalHeadInfo.minCriticalHead;
+        mCriticalHead = mCurrentHead;
+
+        std::vector<Element *> pipeElements;
+        pipeElements = p_solving_strategy->GetPipingElements();
```

Comment:
This could be one line (also note the renaming to pipe_elements): 
```suggestion
        auto pipe_elements = p_solving_strategy->GetPipingElements();
```